// compile with python -m grpc_tools.protoc -I ./ --python_out=. --grpc_python_out=. service.proto

syntax = "proto3";

service Worker {
    rpc ForwardMode(ForwardRequest) returns (ForwardReply) {}
    rpc BackwardMode(BackwardRequest) returns (BackwardReply) {}
    rpc ResetState(EmptyMessage) returns (EmptyMessage) {}
}

message EmptyMessage {}

// The request message containing the user's name.
message ForwardRequest {
    bool inference = 1;
    // bytes systems = 2; // pickle object
    bytes gradients = 2;
    bytes integrators = 3;
    bytes x0 = 4;
    bytes v0 = 5;
    string precision = 6;
    int32 n_frames = 7;
    string key = 8;
}

message ForwardReply {
    bytes du_dls = 1; // numpy array
    bytes energies = 2; // numpy array
    bytes frames = 3;
}

// The response message containing the greetings
message BackwardRequest {
    bytes adjoint_du_dls = 1; // pickle
    string key = 2;
}

message BackwardReply {
    bytes dl_dps = 1; // pickle
    bytes adjoint_x_t = 2; // final frame adjoints
    bytes adjoint_v_t = 3; // final velocities adjoints
}